### PR TITLE
feat(cli): add lacework query preview-source

### PR DESCRIPTION
**Pull request by hazedav**
## Summary
As a user I would like to see a sample event from a specified datasource

## How did you test this change?
Automated integration testing

## Issue
https://lacework.atlassian.net/browse/ALLY-879

## Example
```
❯ lacework-dev query preview-source LW_HE_FILES      
{
  "BATCH_END_TIME": 1644556581058,
  "BATCH_START_TIME": 1644470181058,
  "BLOCK_COUNT": "1760",
  "BLOCK_SIZE": "4096",
  "FILEDATA_HASH": "4f8a3fa663e9756da6b43ae130f29fa8bfb2f8ecf8318f3e6acf897811ed5835",
  "FILE_ACCESSED_TIME": 1632845872000,
  "FILE_CREATED_TIME": 1642126810378,
  "FILE_MODIFIED_TIME": 1632845872000,
  "FILE_NAME": "ldconfig",
  "FILE_PERMISSIONS": "33261",
  "FILE_TYPE": "application/x-executable",
  "HARD_LINK_COUNT": "2",
  "INODE": "8848",
  "IS_LINK": "0",
  "LINK_ABS_DEST_PATH": "",
  "LINK_DEST_PATH": "",
  "METADATA_HASH": "a8958947e742f45b699d8f8c2a9c1c2c9626bc68cd8becfc1a43891a3224c1a7",
  "MID": "4620",
  "OWNER_GID": "0",
  "OWNER_UID": "0",
  "OWNER_USERNAME": "root",
  "PATH": "/usr/sbin/ldconfig",
  "RECORD_CREATED_TIME": 1644499484258,
  "SIZE": "897784"
}
```

### CODE REVIEW

File: cli/cmd/lql_preview.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/lacework/go-sdk/api"
+	"github.com/lacework/go-sdk/lwtime"
+)
+
+var (
+	queryPreviewSourceCmd = &cobra.Command{
+		Hidden: false,
```
Comment (afiune) [inline cli/cmd/lql_preview.go:35]: This is the default, no need to add this line:
```suggestion
```

File: cli/cmd/lql_preview.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		executeQuery := api.ExecuteQueryRequest{
+			Query: api.ExecuteQuery{
+				QueryText: fmt.Sprintf(
+					queryPreviewSourceTemplate, args[0], strings.Join(returns, ",")),
+			},
+			Arguments: queryArgs,
+		}
+		if args[0] == "CloudTrailRawEvents" {
+			executeQuery.Query.EvaluatorID = "Cloudtrail"
+		}
```
Comment (afiune) [inline cli/cmd/lql_preview.go:101]: nit: we don't have to check this for all attempts, we should move this if outside the for loop.

File: cli/cmd/lql_preview.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	if len(returns) == 0 {
+		return errors.New("unable to parse datasource schema")
+	}
+
+	// initialize time attempts
+	timeAttempts := []map[string]string{
+		map[string]string{"start": "-24h", "end": "now"},
+		map[string]string{"start": "-7d", "end": "-24h"},
+		map[string]string{"start": "-30d", "end": "-7d"},
```
Comment (afiune) [inline cli/cmd/lql_preview.go:73]: nit: you don't need to specify the type for every single slice:
```suggestion
		{"start": "-24h", "end": "now"},
		{"start": "-7d", "end": "-24h"},
		{"start": "-30d", "end": "-7d"},
```

File: cli/cmd/lql_preview.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			return errors.Wrap(err, "unable to preview datasource")
+		}
+		if s := reflect.ValueOf(data); s.Kind() != reflect.Slice {
+			err = errors.New("preview results data is not a slice")
+			return errors.Wrap(err, "unable to preview datasource")
+		}
+		if len(data.([]interface{})) == 0 {
+			continue
+		}
+		return cli.OutputJSON(data.([]interface{})[0])
```
Comment (afiune) [inline cli/cmd/lql_preview.go:125]: Using reflect sounds like too much, we could do the same with a simple "try casting" like this:
```suggestion
		if slice, ok := data.([]interface{}); ok {
			if len(slice) == 0 {
				continue
			}
			return cli.OutputJSON(slice[0])
		} else {
			err = errors.New("preview results data is not a slice")
			return errors.Wrap(err, "unable to preview datasource")
		}
```

**Review by afiune — CHANGES_REQUESTED** [review]
After these modifications we should be good to merge! 💯

File: cli/cmd/lql_preview.go (hunk 1 above)
Comment (hazedav) [reply cli/cmd/lql_preview.go:35]: Fixed.

File: cli/cmd/lql_preview.go (hunk 2 above)
Comment (hazedav) [reply cli/cmd/lql_preview.go:101]: Fixed.

File: cli/cmd/lql_preview.go (hunk 4 above)
Comment (hazedav) [reply cli/cmd/lql_preview.go:125]: Fixed.

**Review by afiune — APPROVED** [review]
tenor-79773968

**Comment by afiune** [comment]
@hazedav you have a CI failure:
```
=== Failed                                                                                                                                                                                                                       
=== FAIL: integration TestQueryPreviewNoSuchSource (0.48s)                                                                                                                                                                       
    lql_preview_test.go:51:                                                                                                                                                                                                      
                Error Trace:    lql_preview_test.go:51                                                                                                                                                                           
                Error:          "Usage:\n  lacework query preview-source <datasource_id> [flags]\n\nFlags:\n  -h, --help   help for preview-source\n\nGlobal Flags:\n  -a, --account string      account subdomain of URL (i.e. <
ACCOUNT>.lacework.net)\n  -k, --api_key string      access key id\n  -s, --api_secret string   secret access key\n      --api_token string    access token (replaces the use of api_key and api_secret)\n      --debug           
    turn on debug logging\n      --json                switch commands output from human-readable to json format\n      --nocache             turn off caching\n      --nocolor             turn off colors\n      --noninteracti
ve      turn off interactive mode (disable spinners, prompts, etc.)\n      --organization        access organization level data sets (org admins only)\n  -p, --profile string      switch between profiles configured at ~/.lace
work.toml\n      --subaccount string   sub-account name inside your organization (org admins only)\n\nERROR unable to retrieve data source: \n  [GET] https://customerdemo.lacework.net/api/v2/Datasources/NoSuchSource\n  [404] 
Datasource NoSuchSource not found\n" does not contain "unable to retrieve datasource"                                                                                                                                            
                Test:           TestQueryPreviewNoSuchSource                                                                                                                                                                     
                                                                                                                                                                                                                                 
DONE 195 tests, 1 failure in 395.363s                                                                                                                                                                                            
make: *** [Makefile:57: integration-only] Error 1     
```

A funny one! 😂

**Pull request merged**